editor/src/components/canvas/controls/new-canvas-controls.tsx

```diff
@@ -269,15 +309,173 @@ const NewCanvasControlsInner = (props: NewCanvasControlsInnerProps) => {
+    }
+    if (editorMode.type !== 'select') {
+      return false
+    }
+    return true
+  }, [editorMode, selectionArea])
+
+  const isValidMouseEventForSelectionArea = React.useCallback(
+    (e: React.MouseEvent): boolean => {
+      if (mousePoint == null) {
+        return false
+      }
+      return e.button === 0 && !(e.shiftKey || e.metaKey || e.ctrlKey || e.altKey)
+    },
+    [mousePoint],
+  )
+
+  const selectionAreaCanvasRect: CanvasRectangle | null = React.useMemo(() => {
+    if (selectionArea == null || !canSelectArea) {
+      return null
+    }
+    return rectangleFromTLBR(
+      getCanvasPoint(selectionArea.x, selectionArea.y),
+      getCanvasPoint(selectionArea.x + selectionArea.width, selectionArea.y + selectionArea.height),
+    )
+  }, [selectionArea, getCanvasPoint, canSelectArea])
+
+  const onMouseUp = React.useCallback(
+    (e: React.MouseEvent<HTMLDivElement>) => {
+      setSelectionAreaStart(null)
+      setSelectionAreaRectangle(null)
+
+      const isSelectingArea = isSelectModeWithArea(editorMode)
+      if (isSelectingArea) {
+        let actions: EditorAction[] = [switchEditorMode(EditorModes.selectMode())]
+        if (
+          selectionAreaStart != null &&
+          localHighlightedViews.length > 0 &&
+          isValidMouseEventForSelectionArea(e)
+        ) {
+          actions.push(selectComponents(localHighlightedViews, false))
+        }
+        dispatch(actions)
+        setLocalHighlightedViews([])
+      }
+
+      selectModeHooks.onMouseUp(e)
+    },
+    [
+      dispatch,
+      setSelectionAreaRectangle,
+      selectionAreaStart,
+      localHighlightedViews,
+      isValidMouseEventForSelectionArea,
+      editorMode,
+      selectModeHooks,
+      setLocalHighlightedViews,
+    ],
+  )
+
+  const onMouseMove = React.useCallback(
+    (e: React.MouseEvent<HTMLDivElement>) => {
+      setMousePoint(canvasPoint({ x: e.clientX, y: e.clientY }))
+
+      setSelectionAreaRectangle(
+        getSelectionAreaRenderedRect(selectionArea, ref.current?.getBoundingClientRect() ?? null),
+      )
+
+      if (selectionAreaCanvasRect != null) {
+        const elementsUnderSelectionArea = getElementsUnderSelectionArea(
+          storeRef.current.jsxMetadata,
+          selectionAreaCanvasRect,
+        )
+        if (elementsUnderSelectionArea != null) {
+          setLocalHighlightedViews(elementsUnderSelectionArea)
+        }
+      }
+
+      selectModeHooks.onMouseMove(e)
+    },
+    [
+      storeRef,
+      selectionAreaCanvasRect,
+      selectModeHooks,
+      setLocalHighlightedViews,
+      setSelectionAreaRectangle,
+      selectionArea,
+    ],
+  )
+
+  const onMouseDown = React.useCallback(
+    (e: React.MouseEvent<HTMLDivElement>) => {
+      const possibleElementsUnderMouse = getPossibleElementsUnderMouse(
+        storeRef.current.jsxMetadata,
+        localSelectedViews,
+      )
+      if (
+        !possibleElementsUnderMouse.some(elementIsUnderMouse(mousePointOnCanvas)) &&
```

Comment:
I think these two are essentially checking if there are any `highlightedViews`, since a mousedown event on one of those would select it. If that's the case we should just be checking that, meaning these two functions could be deleted.

Reply:
yes and no - highlighted views are updated when there is no selection, but if there is a selection, moving the mouse over a selected element won't highlight it again (and I believe this is the correct behaviour), so this would fail and prevent the user from being able to move the selected elements (it would restart a new selection area)

Reply:
In that case we could check if there is a selected element under the cursor too?

Reply:
like https://github.com/concrete-utopia/utopia/pull/3750/commits/b5153ca3c31dfa14e5b708a8a0195d0b50bbfbee ?